GroupingToMatrix: Type fix to resolve `localeCompare` error

alexjonspencer1 opened the pull request:
**What does this PR do?**

This is a bugfix PR that resolves a type mismatch causing an error with the `localeCompare` JavaScript string method. Field types for the "Grouping to Matrix" transformation were being statically hardcoded, resulting in calling the `stringComparer` with number types! 🚫

In `groupingToMatrix.ts`, I am now assigning the field `type` based on a dynamic value, instead of the problematic behavior of harcoding it to `string`.

This PR also has some extra stuffz: replacing `localeCompare` with `Intl.Collator` (so much speed 🏃💨) and refactoring to some newer Javascript standards (goodbye for loops).

#### Before

<img width="1376" alt="image" src="https://github.com/user-attachments/assets/3fedddb7-7ec8-4c99-aab2-384ee93aa1e3" />

#### After

<img width="1371" alt="image" src="https://github.com/user-attachments/assets/e218175c-1407-4b6c-85ed-56412f510a1c" />

Fixes https://github.com/grafana/grafana/issues/81440

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

alexjonspencer1 committed:
draft: coerce types to string to prevent localCompare bug

alexjonspencer1 committed:
chore: update with comment for draft

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=98710) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=98710) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=98710) it.</sub>

alexjonspencer1 committed:
chore: cleanup and change fix

alexjonspencer1 commented on packages/grafana-data/src/field/fieldComparers.ts:
```diff
@@ -55,7 +55,11 @@ const stringComparer = (a: string, b: string): number => {
     return falsyComparer(a, b);
   }
-  return a.localeCompare(b);
+
```
Not the root cause of the issue anymore - only strings are being passed to the `stringComparer` 😀, BUT taking the opportunity to change the comparer from `localeCompare` to `Intl.Collator`

alexjonspencer1 commented on packages/grafana-data/src/transformations/transformers/groupingToMatrix.ts:
```diff
@@ -90,33 +86,32 @@ export const groupingToMatrixTransformer: DataTransformerInfo<GroupingToMatrixTr
             name: rowColumnField,
             values: rowValues,
-            type: FieldType.string,
+            type: keyRowField.type,
```
This is the actual fix here!

leeoniya commented on packages/grafana-data/src/field/fieldComparers.ts:
```diff
@@ -55,7 +55,11 @@ const stringComparer = (a: string, b: string): number => {
+
+  // Using the Intl.Collator object compare method results in much faster string sorting
+  // (estimated 2x as fast) than .localeCompare
+  const compare = new Intl.Collator('en', { sensitivity: 'base' }).compare;
```
the trick is to pull this out of the comparison loop. you want to only instantiate this one time globally.

alexjonspencer1 committed:
fix: instantiate compare outside of functions scope

alexjonspencer1 committed:
chore: update tests!

alexjonspencer1 commented on packages/grafana-data/src/transformations/transformers/groupingToMatrix.test.ts:
```diff
@@ -31,7 +31,7 @@ describe('Grouping to Matrix', () => {
         {
           name: 'Time\\Time',
-          type: FieldType.string,
+          type: FieldType.time,
```
Update tests to correct type which is derived from data, not hardcoded to `string`. Type `time` will call the `timeComparer` field compare function, which handles types 🚀

alexjonspencer1 committed:
fix: make the linter happy again

leeoniya commented on packages/grafana-data/src/transformations/transformers/groupingToMatrix.ts:
```diff
@@ -146,10 +141,7 @@ export const groupingToMatrixTransformer: DataTransformerInfo<GroupingToMatrixTr
-  for (let index = 0; index < values.length; index++) {
-    unique.add(values[index]);
-  }
+  values.forEach((value) => unique.add(value));
```
i dont think you need this loop. you can just do `new Set(values)`

alexjonspencer1 replied:
Updated!

alexjonspencer1 committed:
fix: remove unnecessary loop to generate array

leeoniya commented:
sorry, couple more notes :)

typically, `forEach` is slower than `for` loops. it's not noticeable for 100 values, but can be for 10k values. i think switching to `forEach` here isnt really necessary and makes the diff larger.

additionally, reordering properties alphabetically and insetting empty newlines is maybe something your editor is configured to do? it's best to disable these and keep the diff focused on the functionally relevant changes and easier to review :)

![image](https://github.com/user-attachments/assets/2551bf1c-c27a-4b80-95a9-f30c50098e5c)

alexjonspencer1 committed:
chore: revert forEach to for loops, other changes also

alexjonspencer1 committed:
chore: reduce diff

alexjonspencer1 committed:
chore: one more revert

alexjonspencer1 commented:
> sorry, couple more notes :)
> 
> typically, `forEach` is slower than `for` loops. it's not noticeable for 100 values, but can be for 10k values. i think switching to `forEach` here isnt really necessary and makes the diff larger.
> 
> additionally, reordering properties alphabetically and inserting empty newlines is maybe something your editor is configured to do? it's best to disable these and keep the diff focused on the functionally relevant changes, making it easier to review :)
> 
> ![image](https://private-user-images.githubusercontent.com/43234/405046416-2551bf1c-c27a-4b80-95a9-f30c50098e5c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc0MjcxNjMsIm5iZiI6MTczNzQyNjg2MywicGF0aCI6Ii80MzIzNC80MDUwNDY0MTYtMjU1MWJmMWMtYzI3YS00YjgwLTk1YTktZjMwYzUwMDk4ZTVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIxVDAyMzQyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4ZGUyZGEzZTYwMzk0ZDk5MDcxNTQzNzg5OTBlY2I3Y2JkZjIyYWExYmMyY2NhODlkOGY3MGE4MTAwZTZhN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.TaWOUdq4Cpy1Ey_nIULb7MiMsrdhHHGfpJXe4dCvaCs)

No problem! Make sense for the `forEach` vs `for` loops, I've updated the PR 🚀 

Sounds good on the line breaks + alphabetizing. That was the style preference at my former employer (old habits die hard),  but also understand it generates some diff overhead 😆 I've reverted those changes as well.

drew08t commented on packages/grafana-data/src/transformations/transformers/groupingToMatrix.ts:
```diff
@@ -145,12 +145,7 @@ export const groupingToMatrixTransformer: DataTransformerInfo<GroupingToMatrixTr
 };
 
 function uniqueValues<T>(values: T[]): T[] {
-  const unique = new Set<T>();
-
-  for (let index = 0; index < values.length; index++) {
-    unique.add(values[index]);
-  }
-
```
Nice!

drew08t reviewed (approved):
LGTM and works as expected! 🚀

drew08t committed:
Merge branch 'main' into alexjonspencer1/grouping-matrix-locale-compare-bug
